metadata/src/audio/item.rs

```diff
@@ -16,98 +17,255 @@ use librespot_core::{
 
 pub type AudioItemResult = Result<AudioItem, Error>;
 
-// A wrapper with fields the player needs
 #[derive(Debug, Clone)]
-pub struct AudioItem {
-    pub id: SpotifyId,
-    pub spotify_uri: String,
-    pub files: AudioFiles,
-    pub name: String,
-    pub duration: i32,
-    pub availability: AudioItemAvailability,
-    pub alternatives: Option<Tracks>,
-    pub is_explicit: bool,
+pub enum AudioItem {
+    Track(Box<Track>),
+    Episode(Box<Episode>),
 }
 
 impl AudioItem {
     pub async fn get_file(session: &Session, id: SpotifyId) -> AudioItemResult {
-        match id.item_type {
-            SpotifyItemType::Track => Track::get_audio_item(session, id).await,
-            SpotifyItemType::Episode => Episode::get_audio_item(session, id).await,
-            _ => Err(Error::unavailable(MetadataError::NonPlayable)),
+        Ok(match id.item_type {
+            SpotifyItemType::Track => AudioItem::Track(Box::new(Track::get(session, &id).await?)),
+            SpotifyItemType::Episode => {
+                AudioItem::Episode(Box::new(Episode::get(session, &id).await?))
+            }
+            _ => return Err(Error::unavailable(MetadataError::NonPlayable)),
+        })
+    }
+
+    pub fn id(&self) -> SpotifyId {
+        match self {
+            AudioItem::Track(t) => t.id,
+            AudioItem::Episode(e) => e.id,
+        }
+    }
+
+    pub fn track(&self) -> Option<Track> {
+        match self {
+            AudioItem::Track(t) => Some(*t.clone()),
+            AudioItem::Episode(_) => None,
+        }
+    }
+
+    pub fn episode(&self) -> Option<Episode> {
+        match self {
+            AudioItem::Track(_) => None,
+            AudioItem::Episode(e) => Some(*e.clone()),
         }
     }
-}
 
-#[async_trait]
-pub trait InnerAudioItem {
-    async fn get_audio_item(session: &Session, id: SpotifyId) -> AudioItemResult;
-
-    fn allowed_for_user(
-        user_data: &UserData,
-        restrictions: &Restrictions,
-    ) -> AudioItemAvailability {
-        let country = &user_data.country;
-        let user_catalogue = match user_data.attributes.get("catalogue") {
-            Some(catalogue) => catalogue,
-            None => "premium",
+    pub fn spotify_item_type(&self) -> SpotifyItemType {
+        self.id().item_type
+    }
+
+    pub fn spotify_uri(&self) -> Result<String, Error> {
+        self.id().to_uri()
+    }
+
+    pub fn name(&self) -> String {
+        match self {
+            AudioItem::Track(t) => t.name.clone(),
+            AudioItem::Episode(e) => e.name.clone(),
+        }
+    }
+
+    pub fn cover(&self, session: &Session) -> Vec<String> {
+        let image_url = session
+            .get_user_attribute("image-url")
```

Comment:
Better to use `session.spclient().get_image()` here which encapsulates this behavior.

If you want you may improve the `get_image` and (I believe) `get_head_file` functions there to not return `Err` when the attribute cannot be found, but fall back to the known defaults we are now getting from the product info.

Reply:
Can do on the 1st thing maybe on the next I'll see what I can do.

Reply:
`session.spclient().get_image()` is an asyc function that return bytes.

Reply:
Oh yeah :man_facepalming: maybe you can extract that `get_user_attribute` out of it so we don't have two copies of that around. Then have `get_image` call that too.

Reply:
I'll see what I can do.